lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -356,18 +354,24 @@ static function (Query\Parameter $parameter) use ($key): bool {
      */
     public function setParameters($parameters)
     {
-        if (is_array($parameters)) {
-            /** @psalm-var ArrayCollection<int, Parameter> $parameterCollection */
-            $parameterCollection = new ArrayCollection();
-
-            foreach ($parameters as $key => $value) {
-                $parameterCollection->add(new Parameter($key, $value));
-            }
+        if (! is_array($parameters)) {
+            Deprecation::trigger(
+                'doctrine/orm',
+                'https://github.com/doctrine/orm/pull/9816',
+                'Passing an ArrayCollection to %s is deprecated.',
+                __METHOD__
+            );
+            $this->parameters = $parameters;
 
-            $parameters = $parameterCollection;
+            return $this;
         }
 
-        $this->parameters = $parameters;
+        $this->parameters = new ArrayCollection();
```

Comment:
In other methods you're using `$parameters->toArray();` on an `ArrayCollection` to internally get rid of it. Here you explicitly set `$this->parameters` with an  `ArrayCollection`. Even though this PR is about deprecating  `ArrayCollection` as an injectable parameter type, shouldn't this type be prevented at all? Otherwise `getParameters()` would still return an  `ArrayCollection`, which is kind of unexpected for a deprecated type.

Reply:
If `getParameters()` returns something else, we will have a BC-break though, won't we?

Reply:
Ah, right. The others aren't written into `$this->parameters`.